libs/mixer/melody.ts

```diff
@@ -110,6 +110,60 @@ namespace music {
 
     let playToneID = 0
 
+    /**
+     * Play a melody from the melody editor.
+     * @param melody
+     * @param tempo
+     */
+    //% block="play melody $melody at tempo $tempo|(bpm)" blockId=playMelodyFromEditor
+    //% blockNamespace=music
+    //% weight=75 blockGap=8
+    //% group="Melody"
+    //% melody.shadow="melody_editor"
+    //% tempo.min=40 tempo.max=500
+    //% tempo.defl=120
+    export function playMelodyFromEditor(melody: string, tempo: number) {
+        // add tempo to string so it is reflected in simulation
+        // creates format like "C5-174 B4 A G F E D C "
+        let formattedMelody = melody.substr(0, melody.indexOf(' ')) + "-" + tempo + melody.substr(melody.indexOf(' '), melody.length);
+        const song = new Melody(formattedMelody);
+        song.playUntilDone();
+    }
+
+
+    /**
+     * Create a melody with the melody editor.
+     * @param melody
+     */
+    //% block="$melody" blockId=melody_editor
+    //% blockNamespace=music
+    //% blockHidden = true
+    //% weight=75 blockGap=8
+    //% group="Melody" duplicateShadowOnDrag
+    //% melody.fieldEditor="melody"
+    //% melody.fieldOptions.decompileLiterals=true
+    //% melody.fieldOptions.decompileIndirectFixedInstances="true"
+    //% melody.fieldOptions.onParentBlock="true"
+    export function melodyEditor(melody: string): string {
+        let notes: string[] = melody.split(" ");
```

Comment:
Why not just move this parsing code into the other function? Typically when we have a function for a field editor we just have it do nothing and put all of the logic in the actual API. We have a special comment annotation `SHIM=TD_ID` which will make it so that the blocks compiler/decompiler will ignore this function and not emit it.

Reply:
Thanks, it looks much cleaner now!